server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -143,8 +158,19 @@ void run() {
 
                 logger.debug("Reconciliation is complete");
 
-                desiredBalanceMetrics.updateMetrics(allocationStats, desiredBalance.weightsPerNode());
+                updateDesireBalanceMetrics(allocationStats);
+            }
+        }
+
+        private void updateDesireBalanceMetrics(AllocationStats allocationStats) {
+            var stats = statsProvider.stats(desiredBalance);
+            Map<DiscoveryNode, NodeAllocationStats> nodeAllocationStats = new HashMap<>(stats.size());
+            for (var entry : stats.entrySet()) {
+                var node = allocation.nodes().get(entry.getKey());
+                assert node != null;
+                nodeAllocationStats.put(node, entry.getValue());
             }
+            desiredBalanceMetrics.updateMetrics(allocationStats, desiredBalance.weightsPerNode(), nodeAllocationStats);
```

Comment:
I do not think I am following.

New stats claim they represent the current state of the nodes and are populated from `allocationStatsPerNodeRef` that is set from `nodeAllocationStats` or the last argument here.
In its turn the last argument is effectively copied from `stats` entries that represent the desired state.

How is current stats different from desired stats?

Reply:
It is just to manage the dependencies to be able to reuse `AllocationStatsService`. There, we used to pass the entire Allocator only to get the desired balance out of it, which seems to be used for calculating undesired shard count. (git tells me it's your code actually!)

Reply:
... so if you find a different approach to reuse that code more useful/readable, or see some comments missing or renames, please let me know.

Reply:
I see. That component was originally created to expose stats via endpoint and was not originally used in reconciler.

The first one (minor) is that it accesses the cluster state via `clusterService.state()` and not from `allocation`. They should be the same as this is happening during cluster state update computation, but that is not immediately obvious.

The second one (more substantial) is the set of metric it provides. It does not produce the same metrics as we have for computed balance. This might introduce confusing while comparing them and opens a possibility of publishing different metrics under similar names. Also since this service is used in public API, we will be limited in how we can change it. This might become relevant as we discuss the imbalance measurements and experiment with different balancing functions.

Based on this I suggest not to rely on it and instead implement an apm specific service that is used in both computer and reconciler.

Reply:
> The first one (minor) is that it accesses the cluster state via clusterService.state()

I can change that and make it accept also a state. 

> It does not produce the same metrics as we have for computed balance. 

Not sure I follow. The three node-level metrics from the desired balance that we currently have are `shard_count`, `write_load` and `disk_usage`. We have in this PR those three calculated based on current cluster state. There are two more here, that as I mentioned I chose to export, but we can drop them.

> Also since this service is used in public API, we will be limited in how we can change it.

That code is now refactored into its own class. For now we're reusing it because it provides what we need. If we need to later, we can duplicate it, extend it or whatever is appropriate based on what we need.

Reply:
We discussed this thread offline,

Re
>It does not produce the same metrics as we have for computed balance.

It misses weights, but we discussed that they are not necessary for now.
We also discussed the possibility of weight computations diverging. This aspect could be addressed in future when this problem arises.